src/util/request-builder.js

```diff
@@ -60,32 +79,84 @@ export default class RequestBuilder {
    * @returns {Promise.<any>} Service response as promise
    */
   doHttpRequest(): Promise<any> {
-    if (!this.url) {
-      throw new Error('serviceUrl is mandatory for request builder');
-    }
-    let request = new XMLHttpRequest();
     return new Promise((resolve, reject) => {
-      request.onreadystatechange = function() {
-        if (request.readyState === 4) {
-          if (request.status === 200) {
-            let jsonResponse;
-            try {
-              jsonResponse = JSON.parse(request.responseText);
-            } catch (e) {
-              return reject(`${e.message}, ${request.responseText}`);
-            }
-            if (jsonResponse && typeof jsonResponse === 'object' && jsonResponse.code && jsonResponse.message) reject(jsonResponse);
-            else resolve(jsonResponse);
-          } else {
-            reject(request.responseText);
+      if (!this.url) {
+        return reject(
+          new Error(Error.Severity.CRITICAL, Error.Category.NETWORK, Error.Code.MALFORMED_DATA_URI, {
+            url: this.url
+          })
+        );
+      }
+      return this.createXHR(resolve, reject);
+    });
+  }
+
+  createXHR(resolve: Function, reject: Function): void {
+    let request = new XMLHttpRequest();
+    request.onreadystatechange = () => {
+      if (request.readyState === 4) {
+        if (request.status === 200) {
+          let jsonResponse;
+          try {
+            jsonResponse = JSON.parse(request.responseText);
```

Comment:
move the entire parse logic to the level above, as failure of parsing is related to the parser, not the network layer.
Error should be parse error.
So network reports success => we create MultiRequestResult object that validates the data and set success=false if it is invalid => then fetchData receives this, checks for success true/false and if false then emits a bad_server_response error